### PR TITLE
Improve persistent kv store app

### PR DESCRIPTION
**Pull request by shotonoff**
## Issue being fixed or feature implemented

Investigated that the test `TestReactorValidatorSetChanges` is unstable, because of  updating of abci app is inconsistent.
Using single transaction to update the whole entity fixes this issue

## What was done?

Added the ability to update the set of validators in PersistentKVStoreApplication in single transaction

## How Has This Been Tested?

Unit-test

## Breaking Changes

N/A

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: privval/file.go [hunk 1]
```diff
@@ -609,6 +608,19 @@ func (pv *FilePV) UpdatePrivateKey(
 	if _, ok := pv.Key.FirstHeightOfQuorums[quorumHash.String()]; !ok {
 		pv.Key.FirstHeightOfQuorums[quorumHash.String()] = strconv.Itoa(int(height))
 	}
+
+	// var hashes = make([]string, 0, len(pv.Key.PrivateKeys))
+	// for hash := range pv.Key.PrivateKeys {
+	// 	hashes = append(hashes, hash[:8])
+	// }
+	// debugging code
+	fmt.Printf(
```
Comment (lklimek) [inline privval/file.go:617]: remove comment and Printf

File: test/e2e/log.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env sh
```
Comment (lklimek) [inline test/e2e/log.sh:1]: remove additional files

File: abci/example/kvstore/persistent_kvstore.go [hunk 3]
```diff
@@ -24,13 +25,16 @@ const (
 	ValidatorThresholdPublicKeyPrefix       string = "tpk"
 	ValidatorSetQuorumHashPrefix            string = "vqh"
 	ValidatorSetQuorumHashChangePrefix      string = "vqh:"
+	ValidatorSetUpdatePrefix                string = "vsu:"
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply abci/example/kvstore/persistent_kvstore.go:28]: cleanup redundant constants

File: abci/example/kvstore/persistent_kvstore.go [hunk 4]
```diff
@@ -282,6 +300,45 @@ func isQuorumHashTx(tx []byte) bool {
 	return strings.HasPrefix(string(tx), ValidatorSetQuorumHashPrefix)
 }
 
+func (app *PersistentKVStoreApplication) execValidatorSet(tx []byte) error {
+	tx = tx[len(ValidatorSetUpdatePrefix):]
+	values := bytes.Split(tx, []byte{'|'})
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply abci/example/kvstore/persistent_kvstore.go:305]: yep, we even can use `json` to encode/decode data
here the objective was to make backward-compatible code

File: internal/consensus/common_test.go [hunk 5]
```diff
@@ -1104,3 +933,121 @@ func signDataIsEqual(v1 *types.Vote, v2 *tmproto.Vote) bool {
 		bytes.Equal(v1.ValidatorProTxHash.Bytes(), v2.GetValidatorProTxHash()) &&
 		v1.ValidatorIndex == v2.GetValidatorIndex()
 }
+
+type stateQuorumManager struct {
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply internal/consensus/common_test.go:937]: yes, it is not necessary, `stateQuorumManager` doesn't change the state

File: internal/consensus/reactor_test.go [hunk 6]
```diff
@@ -723,15 +722,14 @@ func (u *validatorUpdater) removeValidatorsAt(height int64, count int) (*privVal
 func (u *validatorUpdater) updateStatePrivVals(privValUpdate *privValUpdate, height int64) {
 	for i, proTxHash := range privValUpdate.newProTxHashes {
 		j := u.stateIndexMap[proTxHash.String()]
-		u.states[j].mtx.Lock()
-		u.states[j].privValidator.UpdatePrivateKey(
+		priVal := u.states[j].PrivValidator()
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply internal/consensus/reactor_test.go:725]: reading field `privValidator` was moved into `PrivValidator()` method

File: abci/example/kvstore/persistent_kvstore.go [hunk 7]
```diff
@@ -282,6 +300,45 @@ func isQuorumHashTx(tx []byte) bool {
...
+	thresholdPubKey, err := base64.StdEncoding.DecodeString(string(values[0]))
+	if err != nil {
+		return fmt.Errorf("threshold Pubkey (%s) is invalid base64", string(values[0]))
+	}
+	app.updateThresholdPublicKey(types.UpdateThresholdPublicKey(thresholdPubKey))
+	quorumHash, err := base64.StdEncoding.DecodeString(string(values[1]))
+	if err != nil {
+		return fmt.Errorf("quorum Hash (%s) is invalid base64", string(values[1]))
+	}
+	app.updateQuorumHash(types.UpdateQuorumHash(quorumHash))
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply abci/example/kvstore/persistent_kvstore.go:315]: basically, I just copied this code from the previous implementation, 
the methods `updateQuorumHash` and `updateThresholdPublicKey` return `ResponseDeliverTx` thus it should return `error`, so this is a place for improvement
my suggestion, let's postponed this refactoring in another PR

File: abci/example/kvstore/persistent_kvstore.go [hunk 8]
```diff
@@ -282,6 +300,45 @@ func isQuorumHashTx(tx []byte) bool {
...
+			if err != nil {
+				return fmt.Errorf("pubkey is invalid: %w", err)
+			}
+		}
+		// decode the power
+		power, err := strconv.ParseInt(string(vals[2]), 10, 64)
+		if err != nil {
+			return fmt.Errorf("power (%s) is not an int", vals[2])
+		}
+		app.updateValidatorSet(types.UpdateValidator(proTxHash, pubkey, power, ""))
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply abci/example/kvstore/persistent_kvstore.go:337]: answered above

**Review by shotonoff — COMMENTED** [review]
please have a look at the answers

**Review by lklimek — APPROVED** [review]
LGTM :+1:

**Pull request merged**